### PR TITLE
#166265888 User can borrow books without admin authorization

### PR DESCRIPTION
**Pull request by encodedBicoding**
### What does this PR do?
Allows a user to borrow a book without authorization from admin
1. Ensures a user can use the `/reserve` endpoint in the books router to reserve a book
2. A user is only allowed to reserve a book for only 3 days before it can be accessed by other users
#### Description of Task to be completed?
Create two endpoints:
1: POST `/reserve` :- this allows a user to reserve a book they intend to borrow
2: POST `/reserved_books`:- this allows an admin to confirm if a user actually reserved a book and then allows the user borrows the book.
#### How should this be manually tested?
N/A
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#166265888](https://www.pivotaltracker.com/story/show/166265888)
#### Screenshots (if appropriate)
N/A
#### Questions
N/A

### CODE REVIEW

**Review by Hadeneekeh — DISMISSED** [review]

**Comment by Hadeneekeh** [comment]
Screenshots help to understand how a feature works, I can't find any and documentation is missing as well.

File: server/controllers/bookController.js [hunk 1]
```diff
@@ -227,6 +227,54 @@ class BookController {
...
+        timeToExpire: expiryDate
+      });
+
+    return Utils.successStatus(res, 200, 'Book reserved successfully', {
+      id: reservedBook.id,
+      title,
+      isbn,
+      patronId: reservedBook.patronId,
+      collected: reservedBook.collected,
+      Expires: reservedBook.timeToExpire.toDateString(),
```
Comment (emp-daisy) [inline server/controllers/bookController.js]: What do you use the value for?
And why is it in Title case?

Comment (encodedBicoding) [reply server/controllers/bookController.js]: Title case was a mistake. If I get your question, you want to know what I use the value of the expires property for? It's so a reserved book can expire for another user to be able to either reserve or borrow it

Comment (emp-daisy) [reply server/controllers/bookController.js]: But you do not use it when search for a reserved book. So I expected you to find reserved book where ID and not expired... something like that

Comment (encodedBicoding) [reply server/controllers/bookController.js]: In my next PR, i could write a function that removes the book from the reseverdBooks table if it's expired. Will this suffice?

Comment (encodedBicoding) [reply server/controllers/bookController.js]: My initial idea was to not save any unwanted data in the reserved books table. And being able to check for an expired reserved book will mean i save them in the table.

File: server/controllers/bookController.js [hunk 2]
```diff
@@ -227,6 +227,54 @@ class BookController {
...
+    })
+  }
+
+  static async checkBookReservation(req, res) {
+
+    const { patronId, isbn } = req.body;
+
+    const user = await models.reservedBooks.findOne({ where: {
+      [Op.and]: [{ isbn }, { patronId }]
```
Comment (emp-daisy) [inline server/controllers/bookController.js]: I expected to only find books where the expiry date isn't in the past.

Comment (encodedBicoding) [reply server/controllers/bookController.js]: Ok

File: server/controllers/bookController.js [hunk 3]
```diff
@@ -200,6 +200,57 @@ class BookController {
...
+
+  static async reserveBook(req, res) {
+
+    const { loggedinUser } = req;
+
+    const { isbn } = req.query;
+
+    const { title } = await models.Books.findOne({ where: { isbn }});
+
+    let expiryDate = new Date();
```
[the remark this answers is not in the document]
Comment (emp-daisy) [reply server/controllers/bookController.js:239]: Pls note all this

[the remark this answers is not in the document]
Comment (Hadeneekeh) [reply server/controllers/bookController.js:239]: Do you mean the Hound comments?

File: server/middleware/authenticator.js [hunk 4]
```diff
@@ -1,11 +1,14 @@
 /* eslint-disable require-jsdoc */
 import passport from 'passport';
 import util from '../helpers/utilities';
+import sequelize from 'sequelize';
```
[the remark this answers is not in the document]
Comment (emp-daisy) [reply server/middleware/authenticator.js]: Note this

File: server/index.js [hunk 5]
```diff
@@ -32,6 +33,7 @@ app.all('*', (req, res) =>
 );
 
 app.use(ErrorHandler.sendError);
+app.use(Authenticator.deleteReservedIfExpired);
```
Comment (emp-daisy) [inline server/index.js]: When does this run? Everytime?
@Hadeneekeh @encodedBicoding

File: server/controllers/bookController.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (emp-daisy) [reply server/controllers/bookController.js:239]: Yes

File: server/index.js (hunk 5 above)
Comment (encodedBicoding) [reply server/index.js]: Kinda, It should always check for it.

Comment (emp-daisy) [reply server/index.js]: I think it is only necessary when before reserving, collecting or lending a book

**Review by emp-daisy — APPROVED** [review]

**Pull request merged**